crates/roc_std/src/roc_str.rs

```diff
@@ -182,34 +211,91 @@ impl RocStr {
 
             if target_cap > SmallString::CAPACITY {
                 // The requested capacity won't fit in a small string; we need to go big.
-                let mut roc_list = RocList::with_capacity(target_cap);
+                let mut big_string = BigString::with_capacity(target_cap);
 
-                roc_list.extend_from_slice(small_str.as_bytes());
-
-                *self = RocStr(RocStrInner {
-                    heap_allocated: ManuallyDrop::new(roc_list),
+                unsafe {
+                    std::ptr::copy_nonoverlapping(
+                        self.as_bytes().as_ptr(),
+                        big_string.ptr_to_first_elem(),
+                        self.len(),
+                    )
+                };
+
+                big_string.length = self.len();
+                big_string.capacity_or_ref_ptr = target_cap;
+
+                let mut updated = RocStr(RocStrInner {
+                    heap_allocated: ManuallyDrop::new(big_string),
                 });
+
+                mem::swap(self, &mut updated);
+                mem::forget(updated);
             }
         } else {
-            let mut roc_list = unsafe { ManuallyDrop::take(&mut self.0.heap_allocated) };
+            let mut big_string = unsafe { ManuallyDrop::take(&mut self.0.heap_allocated) };
 
-            roc_list.reserve(bytes);
+            big_string.reserve(bytes);
 
             let mut updated = RocStr(RocStrInner {
-                heap_allocated: ManuallyDrop::new(roc_list),
+                heap_allocated: ManuallyDrop::new(big_string),
             });
 
             mem::swap(self, &mut updated);
             mem::forget(updated);
         }
     }
 
+    #[track_caller]
+    pub fn slice_range(&self, range: Range<usize>) -> Self {
+        match self.try_slice_range(range) {
+            Some(x) => x,
+            None => panic!("slice index out of range"),
+        }
+    }
+
+    pub fn try_slice_range(&self, range: Range<usize>) -> Option<Self> {
+        if self.as_str().get(range.start..range.end).is_none() {
+            None
+        } else if range.end - range.start <= SmallString::CAPACITY {
```

Comment:
Probably not too important, but if the original string is a slice or a big string, I wouldn't do this. I would instead create another slice in this case. Skips the memcpy. Was a huge win in roc when spliting a string by spaces for example. But the host probably won't do too much with this, so may not matter.

Reply:
ah, so we cannot rely on small capacities corresponding to a small string. that assumption may be made somewhere.

Reply:
Yeah. It could be a seamless slice that just happens to be small str size. This was large enough perf gains that I switch over from making small strings to staying as seamless slices.